 # 바이크 셋 딥러닝 실습

 ### 딥러닝을 위해 파이썬 3.8 버전을 이용 (Jupyter Notebook 따로 실행 필요)

In [1]:
import sys
import tensorflow as tf
import keras
import matplotlib as mpl
import seaborn as sns
import numpy as np
import sklearn as sk
import pandas as pd
print(sys.version)
print(tf.__version__)
print(keras.__version__)
print(mpl.__version__)
print(sns.__version__)
print(np.__version__)
print(sk.__version__)
print(pd.__version__)


3.8.13 (default, Mar 28 2022, 06:59:08) [MSC v.1916 64 bit (AMD64)]
2.10.0
2.10.0
3.6.1
0.12.0
1.23.3
1.1.2
1.5.0


In [2]:
tr = pd.read_csv('bike_mod_tr.csv')
test = pd.read_csv('bike_mod_test.csv')
sub = pd.read_csv('sampleSubmission.csv')

tr.shape, test.shape, sub.shape

((10886, 19), (6493, 16), (6493, 2))

### 데이터 탐색

In [3]:
tr.columns

Index(['datetime', 'season', 'holiday', 'workingday', 'weather', 'temp',
       'atemp', 'humidity', 'windspeed', 'casual', 'registered', 'count',
       'year', 'month', 'day', 'hour', 'minute', 'second', 'dayofweek'],
      dtype='object')

In [4]:
test.columns

Index(['datetime', 'season', 'holiday', 'workingday', 'weather', 'temp',
       'atemp', 'humidity', 'windspeed', 'year', 'month', 'day', 'dayofweek',
       'hour', 'minute', 'second'],
      dtype='object')

### 입력과 출력 지정

In [39]:
input_col = ['hour', 'temp', 'dayofweek', 'workingday', 'season', 'weather']
labeled_col = ['count']

X = tr[input_col]
y = tr[labeled_col]
X_val = test[input_col]

In [6]:
from sklearn.model_selection import train_test_split

In [40]:
## 자체 학습, 자체 평가
X_train, X_test, y_train, y_test = train_test_split(X, y, 
                                        random_state=0)

In [10]:
from keras.models import Sequential
from keras.layers import Dense

In [45]:
model = Sequential()
model.add(Dense(30,input_dim=6, activation='relu'))
model.add(Dense(15, activation='relu'))
model.add(Dense(15, activation='relu'))
model.add(Dense(1))

In [46]:
model.compile(loss="mean_squared_error", optimizer='rmsprop')
model.fit(X_train, y_train, epochs=300, batch_size=10)

Epoch 1/300
817/817 [==============================] - 1s 823us/step - loss: 29261.6777
Epoch 2/300
817/817 [==============================] - 1s 796us/step - loss: 20212.7012
Epoch 3/300
817/817 [==============================] - 1s 790us/step - loss: 18745.5449
Epoch 4/300
817/817 [==============================] - 1s 799us/step - loss: 18269.5254
Epoch 5/300
817/817 [==============================] - 1s 798us/step - loss: 17974.5781
Epoch 6/300
817/817 [==============================] - 1s 793us/step - loss: 17708.7070
Epoch 7/300
817/817 [==============================] - 1s 797us/step - loss: 17429.1211
Epoch 8/300
817/817 [==============================] - 1s 815us/step - loss: 17131.1973
Epoch 9/300
817/817 [==============================] - 1s 825us/step - loss: 16875.3281
Epoch 10/300
817/817 [==============================] - 1s 807us/step - loss: 16602.4180
Epoch 11/300
817/817 [==============================] - 1s 864us/step - loss: 16567.1348
Epoch 12/300
817/817 [========

817/817 [==============================] - 1s 794us/step - loss: 5874.7876
Epoch 186/300
817/817 [==============================] - 1s 787us/step - loss: 6023.9370
Epoch 187/300
817/817 [==============================] - 1s 789us/step - loss: 5983.9858
Epoch 188/300
817/817 [==============================] - 1s 788us/step - loss: 6001.1919
Epoch 189/300
817/817 [==============================] - 1s 798us/step - loss: 5924.1538
Epoch 190/300
817/817 [==============================] - 1s 792us/step - loss: 5910.7378
Epoch 191/300
817/817 [==============================] - 1s 790us/step - loss: 5990.6504
Epoch 192/300
817/817 [==============================] - 1s 792us/step - loss: 5945.6362
Epoch 193/300
817/817 [==============================] - 1s 790us/step - loss: 5974.4565
Epoch 194/300
817/817 [==============================] - 1s 796us/step - loss: 6000.6245
Epoch 195/300
817/817 [==============================] - 1s 792us/step - loss: 5880.7432
Epoch 196/300
817/817 [============

In [47]:
model.evaluate(X_test,y_test)

86/86 [==============================] - 0s 741us/step - loss: 8344.0596


8344.0595703125

In [ ]:
27214.970703125
27094.81640625

In [31]:
pred = model.predict(X_val)

203/203 [==============================] - 0s 668us/step


In [33]:
sub['count'] = pred

sub.loc[sub['count']<0, 'count'] = 0

In [36]:
sub.to_csv("nnn_sub_2207.csv", index=False)

In [29]:
mse = np.mean(((pred - y_test )**2))
mse

C:\Users\GSENR\anaconda3\envs\tf2x\lib\site-packages\numpy\core\fromnumeric.py:3430: FutureWarning: In a future version, DataFrame.mean(axis=None) will return a scalar mean over the entire DataFrame. To retain the old behavior, use 'frame.mean(axis=0)' or just 'frame.mean()'
  return mean(axis=axis, dtype=dtype, out=out, **kwargs)


count    27214.972962
dtype: float64

In [21]:
((pred - y_test )**2)

count
6638     34.575355
7975     25.600333
5915   1306.696296
8050  34128.040875
5894  12516.804307
...            ...
6111    172.384807
1917  18444.293268
5702  55561.876011
4813  18202.351859
5125   6731.861214

[2722 rows x 1 columns]

 * epochs=20, batch_size=10의 test mse = 27142.536362
 * epochs=100, batch_size=10의 test mse = 27137.786287
 * epochs=50, batch_size=5의 test mse = 27214.972962
 * epochs=300, batch_size=10의 test mse = 27094.81640625